pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -247,3 +262,76 @@ func (a *SharedBudgetAllocation) Release(ctx context.Context) {
 		putPooledBudgetAllocation(a)
 	}
 }
+
+// BudgetFactory creates memory budget for rangefeed according to system
+// settings.
+type BudgetFactory struct {
+	limit              int64
+	feedBytesMon       *mon.BytesMonitor
+	systemFeedBytesMon *mon.BytesMonitor
+
+	metrics *FeedBudgetPoolMetrics
+}
+
+// NewBudgetFactory creates a factory callback that would create RangeFeed
+// memory budget according to system policy.
+func NewBudgetFactory(
+	ctx context.Context,
+	rootMon *mon.BytesMonitor,
+	memoryPoolSize int64,
+	histogramWindowInterval time.Duration,
+) *BudgetFactory {
+	if !useBudgets || rootMon == nil {
+		return nil
+	}
+	feedSizeLimit := int64(float64(memoryPoolSize) * maxFeedFraction)
+
+	metrics := NewFeedBudgetMetrics(histogramWindowInterval)
+	systemRangeMonitor := mon.NewMonitorInheritWithLimit("rangefeed-system-monitor",
+		systemRangeFeedBudget, rootMon)
+	systemRangeMonitor.SetMetrics(metrics.SystemCurBytesCount, metrics.SystemMaxBytesHist)
```

Comment:
Max is a histogram so that we have percentiles as separate charts. I guess that's because we show it in web ui:
<img width="383" alt="Screenshot 2022-02-24 at 10 13 32" src="https://user-images.githubusercontent.com/5124786/155504543-2dc464e2-24e5-4348-af0b-40aa2ecd36c3.png">
Probably becomes obsolete if we mandate deploying prometheus container alongside.

Reply:
Oh, right. We obviously don't have per-rangefeed metrics, so we need a histogram for better resolution. Makes sense now.